Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,177 @@
+        
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        Task<DeviceModel> GetAsync(string id);
+        
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true);
+
+        /// <summary>
+        /// Create or replace a custom device model.
+        /// </summary>
+        Task<DeviceModel> UpsertAsync(DeviceModel deviceModel);
+
+        /// <summary>
+        /// Delete a custom device model.
+        /// </summary>
+        Task DeleteAsync(string id);
+    }
+    public class CustomDeviceModels : ICustomDeviceModels
+    {
+        private const string STORAGE_COLLECTION = "deviceModels";
+        private const string CUSTOMMODEL = "CustomModel";
+
+        private readonly IStorageAdapterClient storage;
+        private readonly ILogger log;
+
+        public CustomDeviceModels(
+            IStorageAdapterClient storage,
+            ILogger logger)
+        {
+            this.storage = storage;
+            this.log = logger;
+        }
+
+        /// <summary>
+        /// Get list of custom device models.
+        /// </summary>
+        public async Task<IEnumerable<DeviceModel>> GetListAsync()
+        {
+            var data = await this.storage.GetAllAsync(STORAGE_COLLECTION);
+            var results = new List<DeviceModel>();
+            foreach (var item in data.Items)
+            {
+                var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+                deviceModel.ETag = item.ETag;
+                deviceModel.Type = CUSTOMMODEL;
+                results.Add(deviceModel);
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                this.log.Error("Device model id cannot be empty!", () => {});
+                throw new InvalidInputException("Device model id cannot be empty! ");
+            }
+
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+
+            if (generateId)
+            {
+                deviceModel.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Create custom device model: ", () => new { deviceModel });
```

Comment:
Should this be "created" here?

Reply:
also, please don't end log statement with `:`

Reply:
Done